csp/handler.go

```diff
@@ -161,16 +208,59 @@ func (csp *BlindCSP) sharedKeyReq(msg *bearerstdapi.BearerStandardAPIdata,
 	if len(pid) != processIDSize {
 		return fmt.Errorf("wrong process id: %x", pid)
 	}
-	resp := Message{}
-	var ok bool
-	if ok, resp.Response = csp.AuthCallback(ctx.Request, req, pid, SignatureTypeSharedKey); ok {
-		resp.SharedKey, err = csp.SharedKey(pid)
-		if err != nil {
-			return err
+	// Auth Step
+	step, err := strconv.Atoi(ctx.URLParam("step"))
+	if err != nil {
+		// For backwards compatibility
+		step = 0
+	}
+
+	var resp types.Message
+	var authResp types.AuthResponse
+	if authResp = csp.callbacks.Auth(
+		ctx.Request,
+		req,
+		pid,
+		types.SignatureTypeSharedKey,
+		step,
+	); authResp.Success {
+		if authResp.AuthToken == nil {
+			resp.SharedKey, err = csp.SharedKey(pid)
+			if err != nil {
+				return err
+			}
 		}
 	} else {
 		return fmt.Errorf("unauthorized")
 	}
+	resp.Response = authResp.Response
+	resp.AuthToken = authResp.AuthToken
+	return ctx.Send(resp.Marshal(), bearerstdapi.HTTPstatusCodeOK)
+}
+
+func (csp *BlindCSP) info(msg *bearerstdapi.BearerStandardAPIdata,
+	ctx *httprouter.HTTPContext) error {
+	if csp.callbacks.Info == nil {
+		return ctx.Send(nil, bearerstdapi.HTTPstatusCodeOK)
+	}
+	resp := csp.callbacks.Info()
+	if resp == nil {
+		return ctx.Send(nil, bearerstdapi.HTTPstatusCodeOK)
+	}
+	return ctx.Send(resp.Marshal(), bearerstdapi.HTTPstatusCodeOK)
+}
+
+func (csp *BlindCSP) indexer(msg *bearerstdapi.BearerStandardAPIdata,
```

Comment:
@p4u  I understand the choice of making the indexer generic per verification handler type, nevertheless IMHO  it goes in contrast to the idea of a CSP that supports multiple verifications in sequence. In such a scenario, a generic indexer would make sense and a separate (per handler call) that provides user info.